### ISSUE TITLE
[BUG] Unicode characters in SLS files cause an exception on Windows minions

### ISSUE THREAD

**Issue by smarsching** [saltstack/salt#62205]
**Description**
A bug causing the wrong encoding to be detected on Windows causes an exception when using certain unicode characters in SLS files.

**Setup**
Assume we have a simple SLS that uses unicode characters in a comment:

`test/unicode.sls`:
```yaml
# The state does not really matter, the issue is caused by the closing quotation mark (”) in this comment.
test_unicode:
  test.succeed_with_changes
```

The bug is triggered when applying this SLS to a minion running on Windows 10 or Windows Server 2022 (and possibly other versions of Windows).
  
**Steps to Reproduce the behavior**
Create the SLS listed above on the Salt master and run
```
salt windows-minion.example.com state.apply test.unicode
```

This leads to the following exception:

```
windows-minion.example.com:
    Data failed to compile:
----------
    Rendering SLS test.unicode failed, render error: 'charmap' codec can't decode byte 0x9d in position 89: character maps to <undefined>
Traceback (most recent call last):
  File "C:\Program Files\Salt Project\Salt\bin\lib\site-packages\salt-3004.1-py3.8.egg\salt\state.py", line 4025, in render_state
    state = compile_template(
  File "C:\Program Files\Salt Project\Salt\bin\lib\site-packages\salt-3004.1-py3.8.egg\salt\template.py", line 82, in compile_template
    render_pipe = template_shebang(
  File "C:\Program Files\Salt Project\Salt\bin\lib\site-packages\salt-3004.1-py3.8.egg\salt\template.py", line 179, in template_shebang
    line = salt.utils.stringutils.to_unicode(ifile.readline())
  File "C:\Program Files\Salt Project\Salt\bin\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 89: character maps to <undefined>

ERROR: Minions returned with non-zero exit code
```

**Expected behavior**
Applying the SLS should not result in an exception. For comparison, here is the output when applying the same SLS to a minion running on Linux:

```
linux-minion.example.com:
----------
          ID: test_unicode
    Function: test.succeed_with_changes
      Result: True
     Comment: Success!
     Started: 14:28:16.551031
    Duration: 0.498 ms
     Changes:   
              ----------
              testing:
                  ----------
                  new:
                      Something pretended to change
                  old:
                      Unchanged

Summary for linux-minion.example.com
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:   0.498 ms
```

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.) 

```yaml
Salt Version:
          Salt: 3004.1

Dependency Versions:
          cffi: 1.14.6
      cherrypy: 18.6.1
      dateutil: 2.8.1
     docker-py: Not Installed
         gitdb: 4.0.7
     gitpython: Not Installed
        Jinja2: 2.10.1
       libgit2: Not Installed
      M2Crypto: Not Installed
          Mako: 1.1.4
       msgpack: 0.6.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: 2.20
      pycrypto: Not Installed
  pycryptodome: 3.10.1
        pygit2: Not Installed
        Python: 3.8.8 (tags/v3.8.8:024d805, Feb 19 2021, 13:18:16) [MSC v.1928 64 bit (AMD64)]
  python-gnupg: 0.4.7
        PyYAML: 5.4.1
         PyZMQ: 19.0.0
         smmap: 4.0.0
       timelib: 0.2.4
       Tornado: 4.5.3
           ZMQ: 4.3.2

System Versions:
          dist:
        locale: cp1252
       machine: AMD64
       release: 2022Server
        system: Windows
       version: 2022Server 10.0.20348 SP0 Multiprocessor Free
```
</details>

**Additional context**

**TL;DR**

There is a bug in the default encoding detection on Windows, that could easily be fixed by avoiding the use of `locale.getdefaultlocale()` on this platform.

**Detailed discussion**

I believe that this is caused by a combination of problems:

`salt.template.template_shebang(…)` tries to convert the file contents to the correct locale by passing them to `salt.utils.stringutils.to_unicode(…)`, but the exception is triggered before this, when calling `readline()` on the file object returned by `salt.utils.files.fopen(…)`.

I think that this could be solved by opening the file in binary mode (after all, `to_unicode(…)` can handle the decoding of binary data).

However, this raises the interesting question why `readline(…)` raises an exception. From the stack trace, it is clear that it tries to decode the file as Windows CP-1252, when in fact it is encoded with UTF-8.

This can be traced back to `salt.utils.files.fopen(…)` using `__salt_system_encoding__` when no encoding is specified by the calling code. So maybe, `salt.template.template_shebang(…)` should pass `encoding="UTF-8"` when calling `fopen(…)`, but I am not sure whether there is any reasonable case where template files might use an encoding that is not UTF-8.

The other question is why `__salt_system_encoding__` is `cp1252` when any modern Windows system uses UTF-8 by default. In order to explain this, we have to take a look at `__define_global_system_encoding_variable__()` in `salt/__init__.py`:

```py
def __define_global_system_encoding_variable__():
    import sys

    # This is the most trustworthy source of the system encoding, though, if
    # salt is being imported after being daemonized, this information is lost
    # and reset to None
    encoding = None

    if not sys.platform.startswith("win") and sys.stdin is not None:
        # On linux we can rely on sys.stdin for the encoding since it
        # most commonly matches the filesystem encoding. This however
        # does not apply to windows
        encoding = sys.stdin.encoding
```

This comment is interesting, because Python on Windows correctly reports `sys.stdin.encoding` as `utf-8` in my environment, which is the encoding used for the filesystem in any version of Windows since Windows 2000. However, this does not matter so much because in the context of running the minion, `sys.stdin` often won’t be available and so we cannot use it to detect the encoding anyway.

Let’s see what happens next:

```py
    if not encoding:
        # If the system is properly configured this should return a valid
        # encoding. MS Windows has problems with this and reports the wrong
        # encoding
        import locale

        try:
            encoding = locale.getdefaultlocale()[-1]
        except ValueError:
            # A bad locale setting was most likely found:
            #   https://github.com/saltstack/salt/issues/26063
            pass
```

On Windows, `locale.getdefaultlocale()` delegates to `_locale._getdefaultlocale()` which is written in C (the code can be found in `Modules/_localemodule.c` of the CPython sources). This function determines the encoding by calling the `GetACP()` function from the Win32 API. The [documentation](https://docs.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-getacp) for this function effectively says: “Don’t use the return value of this function as the codepage might not really cover all the characters. Use Unicode instead.”

So, we end up using an encoding, even though the comment in the code says that it is most likely wrong on Windows and Microsoft’s documentation (and practical evidence) supports this assumption.

Let’s see what would happen if we skipped the call to `locale.getdefaultlocale()` on Windows by taking a look at the next lines in `__define_global_system_encoding_variable__()`’s  source code:

```py
        # This is now garbage collectable
        del locale
        if not encoding:
            # This is most likely ascii which is not the best but we were
            # unable to find a better encoding. If this fails, we fall all
            # the way back to ascii
            encoding = sys.getdefaultencoding()
```

So, if `locale.getdefaultlocale()` wasn’t used, the return value of `sys.getdefaultencoding()` would be used. Interestingly, this function returns the correct value (`utf-8`) on my Windows system. I think that the comment might refer to older versions of Python (Python 2.x) where the default encoding was ASCII and not UTF-8.

I think the proper way to fix this is skipping `locale.getdefaultlocale()` when running on Windows and directly going to `sys.getdefaultencoding()`. If I read the [source code](https://github.com/python/cpython/blob/main/Python/sysmodule.c) for this function in CPython correctly, this function always returns `utf-8` in Python 3.x, regardless of the platform.

This would effectively mean that we assume UTF-8 when running on Windows, which should be correct for all versions of Windows that are still in use.

I am happy to provide a PR that makes this change (after all, it’s a one-liner), but I first wanted to get some feedback about whether you think that this is the correct way to fix this issue. As written earlier, we could also solve this particular exception by opening the file in binary mode, but fixing the root cause of using the wrong encoding when reading files seems more reasonable to me, because this will most likely affect other parts of Salt as well (for example, I think that some of the issues described in #58462) might be related to this as well.

**Comment by oijkn** [comment]
I confirm that I have the same problem with `file.replace` on a Windows Server 2019.

```
An exception occurred in this state: Traceback (most recent call last):
                  File "c:\salt\bin\lib\site-packages\salt-3004.2-py3.8.egg\salt\state.py", line 2179, in call
                    ret = self.states[cdata["full"]](
                  File "c:\salt\bin\lib\site-packages\salt-3004.2-py3.8.egg\salt\loader\lazy.py", line 149, in __call__
                    return self.loader.run(run_func, *args, **kwargs)
                  File "c:\salt\bin\lib\site-packages\salt-3004.2-py3.8.egg\salt\loader\lazy.py", line 1201, in run
                    return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
                  File "c:\salt\bin\lib\site-packages\salt-3004.2-py3.8.egg\salt\loader\lazy.py", line 1216, in _run_as
                    return _func_or_method(*args, **kwargs)
                  File "c:\salt\bin\lib\site-packages\salt-3004.2-py3.8.egg\salt\loader\lazy.py", line 1249, in wrapper
                    return f(*args, **kwargs)
                  File "C:\salt\bin\Lib\site-packages\salt-3004.2-py3.8.egg\salt\states\file.py", line 5061, in replace
                    changes = __salt__["file.replace"](
                  File "c:\salt\bin\lib\site-packages\salt-3004.2-py3.8.egg\salt\loader\lazy.py", line 149, in __call__
                    return self.loader.run(run_func, *args, **kwargs)
                  File "c:\salt\bin\lib\site-packages\salt-3004.2-py3.8.egg\salt\loader\lazy.py", line 1201, in run
                    return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
                  File "c:\salt\bin\lib\site-packages\salt-3004.2-py3.8.egg\salt\loader\lazy.py", line 1216, in _run_as
                    return _func_or_method(*args, **kwargs)
                  File "c:\salt\bin\lib\site-packages\salt-3004.2-py3.8.egg\salt\modules\file.py", line 2630, in replace
                    w_file.write(salt.utils.stringutils.to_str(result))
                  File "c:\salt\bin\lib\encodings\cp1252.py", line 19, in encode
                    return codecs.charmap_encode(input,self.errors,encoding_table)[0]
                UnicodeEncodeError: 'charmap' codec can't encode character '\ufeff' in position 0: character maps to <undefined>
```

**Comment by twangboy** [comment]
I cannot produce this error on 3006.6. I believe it was fixed with the PR: https://github.com/saltstack/salt/pull/65791 which defaults to utf-8 on all systems.
